### ISSUE TITLE
GravesLSTM.gradient(), not yet implemented

### ISSUE THREAD

**Issue by jadams5** [deeplearning4j/deeplearning4j#516]
When trying to use GravesLSTM in a multilayer network, it throws the following exception:
```
Exception in thread "main" java.lang.UnsupportedOperationException: Not yet implemented
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.gradient(GravesLSTM.java:65)
	at org.deeplearning4j.nn.layers.BaseLayer.gradientAndScore(BaseLayer.java:499)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:115)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.optimize(BaseOptimizer.java:131)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:52)
	at org.deeplearning4j.nn.layers.BaseLayer.fit(BaseLayer.java:493)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.pretrain(MultiLayerNetwork.java:215)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1267)
	at com.tsi.DL4JTest.main(DL4JTest.java:73)
```

**Comment by AlexDBlack** [comment]
Some API changes will be pushed to master soon. I'll do another pass on GravesLSTM then to make sure it works. At present GravesLSTM not working with MultiLayerNetwork is a known issue.
Also your stack trace suggests your configuration might be wrong?
at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.pretrain(MultiLayerNetwork.java:215)
i.e., pretraining is not possible for LSTM.

**Comment by AlexDBlack** [comment]
I recently went through GravesLSTM as part of API changes (soon to be merged to master), and it passes unit tests.
To clarify my previous comment: .gradient() is for unsupervised pretraining (RBMs, autoencoders) vs. backpropGradient() for backprop. So it shouldn't be calling gradient() if configured correctly.
Re-check this after api changes get merged (branch ab_layerbackprop) and double check your config. If you are still having issues with this then reopen the issue.

**Comment by agibsonccc** [comment]
This needs to stay open. Gradient needs to be implemented for the optimizers in all situations.

**Comment by AlexDBlack** [comment]
OK, that's a design issue then we need to discuss. Melanie and I were thinking of setting optimizers to get either backprop gradient or pretrain gradient depending on configuration.
Not sure pretrain makes sense at all with LSTM?

**Comment by agibsonccc** [comment]
This has been fixed.
